reps/2022-04-01-resource-control.md

```diff
@@ -0,0 +1,199 @@
+mount | grep '^cgroup' | awk '{print $1}' | uniq
+```
+
+And you can try to enable cgroup v2:
+```
+grubby --update-kernel=ALL --args="systemd.unified_cgroup_hierarchy=1"
+reboot
+```
+
+Overall, cgroup v2 seems more reasonable and easy to use, but we should also support cgroup v1 because v1 is widely used and has been hard coded to the [OCI](https://opencontainers.org/) standards. You can see this [blog](https://www.redhat.com/sysadmin/fedora-31-control-group-v2) for more.
+
+#### cgroupfs
+The traditional way to manage cgroups is writing the file system directly. It is usually referred to as cgroupfs, like:
+```
+mkdir /sys/fs/cgroup/{worker_id} # Create cgroup
+echo "200000 1000000" > /sys/fs/cgroup/foo/cpu.max # Set cpu quota
+echo {pid} > /sys/fs/cgroup/foo/cgroup.procs # Bind process.
+```
+NOTE: This is an example based on cgroup v2. The commmand lines in cgroup v1 is defferent and incompatible.
+
+We also could use the [libcgroup](https://github.com/libcgroup/libcgroup/blob/main/README) to simplify the implementation. This library support both cgroup v1 and cgroup v2. 
+
+#### systemd
+Systemd, the init daemon of most of linux hosts, is also a cgroup driver which can use cgroup to manage processes. It provide a wrapped way to bind process and cgroup, like:
+```
+systemctl set-property {worker_id}.service CPUQuota=20% # Bind process and cgroup automatically.
+systemctl start {worker_id}.service
+```
+
+NOTE: The entire config options is [here](https://man7.org/linux/man-pages/man5/systemd.resource-control.5.html). And we can also use `StartTransientUnit` to create cgroup with worker process simply. This is a [dbus](https://www.freedesktop.org/wiki/Software/systemd/dbus/) API and there is a [dbus-python](https://dbus.freedesktop.org/doc/dbus-python/) module we can used.
+
+#### Why we should support both cgroupfs and systemd
+The cgroupfs and the systemd are two mainstream ways to manage cgroups. In [container technology](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/), this two ways are also supported. We should also support both of them.
+
+The reason of using cgroupfs:
+- Cgroupfs is a traditional and common way to manage cgroups. If your login user has the privilege of cgroup file system, you can create the cgroups.
+- In some environments, you couldn't use systemd, e.g. in a non-systemd based system or in a container which couldn't access the systemd service.
+
+The reason of using systemd:
+- Systemd is highly recommended by [runc](https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md.) for cgroup v2.
+- If we use cgroupfs in a systemd based system, there will be more than one component which manage the cgroup tree.
+- In the systemd [docs](https://www.freedesktop.org/wiki/Software/systemd/ControlGroupInterface/), we can know that  create/delete cgroups will be unavailable to userspace applications, unless done via systemd's APIs.
+- The systemd has a good abstract API of cgroups. 
+
```

Comment:
So the conclusion is that, we support both cgroupfs and systemd. And systemd will be used when it is available, otherwise we fallback to cgroupfs. Right? can you also add the conclusion to the doc?

Reply:
Maybe we should make it to be a config?

Reply:
Can you clarify the design here? Is it fallback based or config based? If config based, what is the documentation for user here?

Reply:
@simon-mo Yep, I will add this details.